### PR TITLE
Drop redundant shell hook subprocess for `run` subcommand

### LINKED ISSUE #14993: `conda run` could be faster
### Checklist

- [x] I added a descriptive title
- [x] I searched for other speed issues and didn't find a duplicate

### What happened?

`conda run` feels slow to run. A simple `echo hello`:

```
$ time conda run echo hello
hello

conda run echo hello  0.77s user 0.22s system 79% cpu 1.244 total
```

Meanwhile, `micromamba`:

```
$ time micromamba run echo hello
hello
micromamba run echo hello  0.02s user 0.01s system 45% cpu 0.079 total
```

A lot of that is Python vs C++, but looking at the code of `conda run`, I see one obvious opportunity for optimization, and then maybe a controversial one:

* Save one `conda` subprocess:
  * We initialize `conda` in the wrapper script by sourcing [`conda.sh`](https://github.com/conda/conda/blob/main/conda/shell/etc/profile.d/conda.sh) (or equivalent).
  * Then, we use this to call `conda activate`, which starts a new Python process to generate and eval shell code. 
  * However, we know the arguments already, so we could directly call `conda.activate.Activator(["shell.bash", "activate", target_prefix])` when we are generating the wrapper script and paste it there, saving one unnecessary `conda` process startup.
* Spicy 🌶️. I don't think `conda run` needs the plugin system, so it could be special cased along with `activate` and all the `shell.*` subcommands

### Debug

```shell
NA
```

### Conda info

```shell
active environment : None
       user config file : /Users/jrodriguez/.condarc
 populated config files : /Users/jrodriguez/.local/conda/.condarc
                          /Users/jrodriguez/.condarc
          conda version : 25.5.1
    conda-build version : not installed
         python version : 3.12.8.final.0
                 solver : libmamba (default)
       virtual packages : __archspec=1=m1
                          __conda=25.5.1=0
                          __osx=15.5=0
                          __unix=0=0
       base environment : /Users/jrodriguez/.local/conda  (writable)
      conda av data dir : /Users/jrodriguez/.local/conda/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/osx-arm64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /Users/jrodriguez/.local/conda/pkgs
                          /Users/jrodriguez/.conda/pkgs
       envs directories : /Users/jrodriguez/.local/conda/envs
                          /Users/jrodriguez/.conda/envs
               platform : osx-arm64
             user-agent : conda/25.5.1 requests/2.32.3 CPython/3.12.8 Darwin/24.5.0 OSX/15.5 solver/libmamba conda-libmamba-solver/25.4.0 libmambapy/2.3.0
                UID:GID : 501:20
             netrc file : None
           offline mode : False
```

### Conda config

```shell
==> /Users/jrodriguez/.local/conda/.condarc <==
channels:
  - conda-forge

==> /Users/jrodriguez/.condarc <==
show_channel_urls: True
```

### Conda list

```shell
NA
```

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by agriyakhetarpal**
### Description

With this PR, the wrapper script no longer computes or evaluates the shell hook subprocess's output and is deemed redundant. Instead, we rely on the inline activators to set up activation environment variables directly, which are a superset of those the shell hook sets (i.e., the `CONDA_EXE`, `_CONDA_EXE`, `CONDA_PYTHON_EXE`, `_CONDA_ROOT`, `_CE_M`, `_CE_CONDA` variables plus `PATH`, `CONDA_SHLVL`, and `CONDA_PROMPT_MODIFIER`).

This grants us a ~2.89x speedup, i.e., a ~65.39% reduction in mean execution time. I tested with `hyperfine --warmup 10 --runs 100 "conda run -- python --version"`:

#### Before

```
Time (mean ± σ):      1.399 s ±  0.296 s    [User: 0.892 s, System: 0.364 s]
Range (min … max):    1.118 s …  2.621 s    100 runs
```

#### After

```
Time (mean ± σ):     484.1 ms ± 146.2 ms    [User: 308.8 ms, System: 125.6 ms]
Range (min … max):   350.6 ms … 1007.5 ms    100 runs
```

Closes #14993. We discussed the prospect of `conda run` not requiring the plugin system and agreed to converse about it in a new, separate issue: https://github.com/conda/conda/pull/15534#issuecomment-3800305523.

cc: @jaimergp

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](https://github.com/conda/conda/blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [ ] ~Add / update necessary tests?~
- [ ] ~Add / update outdated documentation?~

### CODE REVIEW

**Comment by agriyakhetarpal** [comment]
Re-running CI to get a better estimate for CodSpeed, since this PR is meant to be a performance improvement, not a regression. This PR remains ready for review, otherwise.

**Comment by jezdez** [comment]
Hold on this changes fundamentally how the plugins are loaded this isn't just a bug fix.

**Comment by agriyakhetarpal** [comment]
Yeah, it's more of a performance improvement rather than a bug fix

**Comment by agriyakhetarpal** [comment]
I experimented with another step towards lazily generating parsers in this commit: https://github.com/conda/conda/pull/15672/changes/b02e916ee50e13d3646b8f906f2617c9339c4e4a, but reverted it in a later one since it breaks a few tests. Nevertheless, I do think I should clarify here that the majority of the speedup achieved in this PR has come from dropping the shell hook subprocess, not by wrangling the imports, which have a much lesser effect. 

Happy to hear any concerns, of course. Thanks!

**Comment by agriyakhetarpal** [comment]
> I should clarify here that the majority of the speedup achieved in this PR has come from dropping the shell hook subprocess, not by wrangling the imports, which have a much lesser effect.

To elaborate: if we have the current change (i.e., both changing the import order + removing the shell hook subprocess):

```
Time (mean ± σ):     502.7 ms ± 141.8 ms    [User: 314.7 ms, System: 130.9 ms]
Range (min … max):   354.5 ms … 916.9 ms    100 runs
```

If I drop aee39a9a0eb0c23de0e43a7c84df797045ed43cd locally, i.e., if we keep just the shell hook subprocess removal change in this PR, I get this:

```
Time (mean ± σ):     512.5 ms ± 171.9 ms    [User: 321.3 ms, System: 132.8 ms]
Range (min … max):   360.3 ms … 1297.3 ms    100 runs
```

I tested with `hyperfine --warmup 10 --runs 100 "conda run -- python --version"`. Since there is not much difference in the results from changing the import order beyond purported theoretical improvements, I will revert the change for now. I agree that, since built-in plugins are core functionality as pointed out to me internally, this change is best discussed elsewhere.

File: conda/utils.py [hunk 1]
```diff
@@ -311,35 +311,21 @@ def wrap_subprocess_call(
...
         with Utf8NamedTemporaryFile(mode="w", delete=False) as fh:
             # If any of these calls to the activation hook scripts fail, we want
             # to exit the wrapper immediately and abort `conda run` right away.
             fh.write("set -e\n")
             if dev_mode:
                 from . import CONDA_SOURCE_ROOT
 
-                fh.write(">&2 export PYTHONPATH=" + CONDA_SOURCE_ROOT + "\n")
-            hook_quoted = quote_for_shell(*conda_exe, "shell.posix", "hook", *dev_args)
+                fh.write(f"export PYTHONPATH={CONDA_SOURCE_ROOT}\n")
```
Comment (jaimergp) [inline conda/utils.py]:
```suggestion
                fh.write(f">&2 export PYTHONPATH={CONDA_SOURCE_ROOT}\n")
```

Accidental or intentional removal?

**Comment by jaimergp** [comment]
> With this PR, the wrapper script no longer computes or evaluates the shell hook subprocess's output and is deemed redundant. 

Huh, so basically we missed this bit in the previous optimization work, right? Not sure why Codspeed is not picking up the improvements, though 🤔

**Comment by agriyakhetarpal** [comment]
> > With this PR, the wrapper script no longer computes or evaluates the shell hook subprocess's output and is deemed redundant. 
> 
> Huh, so basically we missed this bit in the previous optimization work, right? Not sure why Codspeed is not picking up the improvements, though 🤔 

Yes, exactly!

I have no clue about CodSpeed, either, but it previously shower regressions, so I can investigate briefly if you'd suggest me to.

**Comment by jaimergp** [comment]
FWIW, my benchmarks confirm the results. Pretty impressive!

```
$ hyperfine --warmup 2 -L branch main,conda-run-misc-speedups --prepare 'git checkout {branch} &> /dev/null' --command-name '{branch}' --export-markdown /var/folders/1b/z374sygs1s96t_xxxv0hr3lr0000gn/T/tmp.hvWVJ09RVM 'python -m conda run -- python -V'

Benchmark 1: main (branch = main)
  Time (mean ± σ):     908.6 ms ± 131.6 ms    [User: 719.3 ms, System: 142.6 ms]
  Range (min … max):   824.6 ms … 1257.3 ms    10 runs
 
Benchmark 2: conda-run-misc-speedups (branch = conda-run-misc-speedups)
  Time (mean ± σ):     327.7 ms ±  59.8 ms    [User: 235.9 ms, System: 53.2 ms]
  Range (min … max):   264.7 ms … 437.9 ms    10 runs
 
Summary
  conda-run-misc-speedups (branch = conda-run-misc-speedups) ran
    2.77 ± 0.65 times faster than main (branch = main)
```

| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `main (branch = main)` | 908.6 ± 131.6 | 824.6 | 1257.3 | 2.77 ± 0.65 |
| `conda-run-misc-speedups (branch = conda-run-misc-speedups)` | 327.7 ± 59.8 | 264.7 | 437.9 | 1.00 |

File: conda/utils.py [hunk 2]
```diff
@@ -311,35 +311,21 @@ def wrap_subprocess_call(
...
             if debug_wrapper_scripts:
                 fh.write(">&2 echo '*** environment before ***'\n>&2 env\n")
-                fh.write(f'>&2 echo "$({hook_quoted})"\n')
-            fh.write(f'eval "$({hook_quoted})"\n')
```
Comment (jezdez) [inline conda/utils.py:339]: Huh, we're completely removing the running of `shell.posix`? I don't understand how that is possible.

**Comment by jezdez** [comment]
Please make sure to wait for Ken and me to finish our reviews

File: conda/utils.py (hunk 2 above)
Comment (agriyakhetarpal) [reply conda/utils.py:339]: Yes! The inline activator sets the variables it was setting, plus a few extra ones that I listed in the description. 

Edit: oh, I assume you're asking why it was removed, and it's because we don't need any interactive features for `conda run`, so shell functions such as `__conda_activate()` or `__conda_hashr()` wouldn't be useful. We call the activator class to generate the activation commands and run the user command, and it saves one subprocess.

Comment (jaimergp) [reply conda/utils.py:339]: This can be considered a follow up to https://github.com/conda/conda/pull/15534, where the activation logic was inlined in the `conda run` script.

Comment (jaimergp) [reply conda/utils.py:339]: In 25.11.0, `conda run python -V` would generate this script:

```
eval "$(/Users/jrodriguez/.local/conda/bin/conda shell.posix hook)"
conda activate  /Users/jrodriguez/.local/conda/envs/conda-dev
python -V
```

With https://github.com/conda/conda/pull/15534 (current `main`), the `conda activate` call was inlined to avoid a subprocess call:

```bash
set -e
eval "$(/Users/jrodriguez/.local/conda/bin/conda shell.posix hook)"
export _CE_M=''
export _CE_CONDA=''
PS1='(conda-dev) '
export PATH='/Users/jrodriguez/.local/conda/envs/conda-dev/bin:/Users/jrodriguez/.local/conda/condabin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/opt/X11/bin:/Users/jrodriguez/noconda/bin:/Users/jrodriguez/.pixi/bin:/Users/jrodriguez/.local/conda/condabin:/Users/jrodriguez/.local/bin:/Users/jrodriguez/.lmstudio/bin:/Users/jrodriguez/.vscode/extensions/ms-python.debugpy-2025.18.0-darwin-arm64/bundled/scripts/noConfigScripts:/Users/jrodriguez/.pixi/bin:/Users/jrodriguez/.local/conda/condabin:/Users/jrodriguez/.local/bin'
export CONDA_SHLVL='1'
export CONDA_PROMPT_MODIFIER='(conda-dev) '
export CONDA_EXE='/Users/jrodriguez/.local/conda/envs/conda-dev/bin/conda'
export _CONDA_EXE='/Users/jrodriguez/.local/conda/envs/conda-dev/bin/conda'
export CONDA_PYTHON_EXE='/Users/jrodriguez/.local/conda/envs/conda-dev/bin/python'
export _CONDA_ROOT='/Users/jrodriguez/.local/conda/envs/conda-dev'
. "/Users/jrodriguez/.local/conda/envs/conda-dev/etc/conda/activate.d/libxml2-split_activate.sh"
set +e
python -V
_CONDA_EXE_RC=$?
exit $_CONDA_EXE_RC
```

This PR removes the leftover `shell.posix hook` call, which was only needed to enable the original `conda activate` call, thus saving one more process:

```bash
set -e
export _CE_M=''
export _CE_CONDA=''
PS1='(conda-dev) '
export PATH='/Users/jrodriguez/.local/conda/envs/conda-dev/bin:/Users/jrodriguez/.local/conda/condabin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/opt/X11/bin:/Users/jrodriguez/noconda/bin:/Users/jrodriguez/.pixi/bin:/Users/jrodriguez/.local/conda/condabin:/Users/jrodriguez/.local/bin:/Users/jrodriguez/.lmstudio/bin:/Users/jrodriguez/.vscode/extensions/ms-python.debugpy-2025.18.0-darwin-arm64/bundled/scripts/noConfigScripts:/Users/jrodriguez/.pixi/bin:/Users/jrodriguez/.local/conda/condabin:/Users/jrodriguez/.local/bin'
export CONDA_SHLVL='1'
export CONDA_PROMPT_MODIFIER='(conda-dev) '
export CONDA_EXE='/Users/jrodriguez/.local/conda/envs/conda-dev/bin/conda'
export _CONDA_EXE='/Users/jrodriguez/.local/conda/envs/conda-dev/bin/conda'
export CONDA_PYTHON_EXE='/Users/jrodriguez/.local/conda/envs/conda-dev/bin/python'
export _CONDA_ROOT='/Users/jrodriguez/.local/conda/envs/conda-dev'
. "/Users/jrodriguez/.local/conda/envs/conda-dev/etc/conda/activate.d/libxml2-split_activate.sh"
set +e
python -V
_CONDA_EXE_RC=$?
exit $_CONDA_EXE_RC
```

Comment (agriyakhetarpal) [reply conda/utils.py:339]: Yes, correct ✅ In #15534, I was going even further by getting the inline activators to avoid changing the shell prompt (we don't need to bother with it for a non-interactive shell) and save some extra time, but that experiment didn't work out. I haven't made that change here to keep this PR narrow in scope.

Comment (jaimergp) [reply conda/utils.py:339]: You could argue that by removing `shell.posix hook` call, the `conda` shell function is no longer defined in the `conda run` script, so if the user depended on that function being present (e.g. manually calling `conda run -- conda activate my-env`), then that'll fail.

I doubt that was a widespread use case, but I did check and somehow it's a [pattern in Dockerfiles](https://github.com/search?q=%2Fconda+run.*%5Cs%2Bconda%5Cs%2B%2F++NOT+is%3Aarchived+NOT+is%3Afork&ref=opensearch&type=code). For example:

```Dockerfile
RUN conda run -n NeuralPLexer conda env update --file NeuralPLexer.yml && conda run -n NeuralPLexer conda clean -afy
``` 

So I think the main point we need to enable this optimization fully is to make sure that the base `conda` is in PATH. In other words, ensure that `$ROOT_PREFIX/condabin` is in PATH, @agriyakhetarpal.

---

Thanks for opening this question @jezdez!

**Comment by agriyakhetarpal** [comment]
Thanks for your investigation, @jaimergp 🚀  I had no clue about these (uncommon, and a bit peculiar, if I can say so) user patterns. I'll mark this as a draft again until I figure out a way to implement https://github.com/conda/conda/pull/15672#discussion_r2788523040

**Comment by agriyakhetarpal** [comment]
(gotta figure out a test for this)

File: conda/utils.py (hunk 1 above)
Comment (jaimergp) [reply conda/utils.py]: Reminder here @agriyakhetarpal

File: conda/utils.py [hunk 3]
```diff
@@ -352,6 +343,11 @@ def wrap_subprocess_call(
 
             fh.write(activate_code)
 
+            # Ensure that nested `conda` commands will work
+            # without the shell hook function.
+            condabin_dir = join(root_prefix, "condabin")
+            fh.write(f'export PATH="{condabin_dir}:$PATH"\n')
```
Comment (jaimergp) [inline conda/utils.py]: Not sure if we should add `condabin` first or last 🤔 Ideally we add it _before_ activating the target environment, so the target environment's `conda` takes precedence (I'm trying to follow the principle of least surprise here, not sure if you agree).

Comment (agriyakhetarpal) [reply conda/utils.py]: Yes, good point, we don't want the base conda to shadow a conda in the target environment.

Comment (agriyakhetarpal) [reply conda/utils.py]: See 97629cf6968061ff376827a06146b88a7717b080

Comment (jaimergp) [reply conda/utils.py]: Thanks, that's reasonable. I would still write it to the script _before_ we inline the activation code, so it behaves even more like the shell.hook logic. i.e., PATH should have:

- PATH entries added by activation
- condabin entry
- whatever was on PATH before that

The script should then do:

```bash
# preamble
set PATH="<condabin>:$PATH"
# inlined activation code
# user command
```

Comment (agriyakhetarpal) [reply conda/utils.py]: Ah, sorry, looks like I misunderstood you in that case. 😅 Let me have a look again!

Comment (agriyakhetarpal) [reply conda/utils.py]: The activator writes an absolute PATH string. So we'll need to somehow inject `condabin` into the process's PATH before the activator runs its `_get_starting_path_list()`. This sounds a bit messy.

Comment (agriyakhetarpal) [reply conda/utils.py]: Could you check 8c357c09b3a6a974c17059e3db7e5539adb63465, @jaimergp? I think it should achieve what we desire, but if you have a better way, LMK!

Comment (jaimergp) [reply conda/utils.py]: Yep, that works, thanks!

File: conda/utils.py (hunk 1 above)
Comment (agriyakhetarpal) [reply conda/utils.py]: Sorry, I missed this comment. This is an intentional removal since it's a no-op. `export` is silent anyway, and there is no benefit in redirecting it to stderr. I assume that this came from someone's old debugging habit? 😄

Comment (kenodegard) [reply conda/utils.py]: looks like a consistency choice 🤷🏼‍♂️

File: tests/cli/test_main_run.py [hunk 4]
```diff
@@ -504,6 +504,36 @@ def test_run_preserves_exit_code_despite_deactivation_failure(
         assert deactivation_marker_file.exists()
 
 
+def test_run_wrapper_has_condabin_on_path(
+    request: pytest.FixtureRequest,
+    tmp_env: TmpEnvFixture,
+    monkeypatch: pytest.MonkeyPatch,
+):
+    """Regression test to ensure condabin is on PATH in the wrapper script.
+    See https://github.com/conda/conda/pull/15672 for context.
```
Comment (kenodegard) [inline tests/cli/test_main_run.py]: This isn't a regression?

Comment (kenodegard) [reply tests/cli/test_main_run.py]:
```suggestion
    """Test to ensure condabin is on PATH in the wrapper script.
    See https://github.com/conda/conda/pull/15672 for context.
```

File: tests/cli/test_main_run.py [hunk 5]
```diff
@@ -504,6 +504,36 @@ def test_run_preserves_exit_code_despite_deactivation_failure(
...
+
+    Without the shell hook, the conda shell functions are not defined,
+    so nested conda commands rely on the conda executable in ``condabin``
+    being discoverable via PATH.
```
Comment (kenodegard) [inline tests/cli/test_main_run.py:517]: Very small corner case but this also means `conda [de]activate` are no longer possible within a `conda run` command.

Currently the following passes (not that we can do much with it):

```
$ conda run -- conda activate scratch
$ echo $?
0
```

With this PR:

```
$ conda run -- conda activate scratch

CondaError: Run 'conda init' before 'conda activate'

ERROR conda.cli.main_run:execute(148): `conda run conda activate scratch` failed. (See above for error)
$ echo $?
1
```

File: tests/cli/test_main_run.py (hunk 4 above)
Comment (agriyakhetarpal) [reply tests/cli/test_main_run.py]: Thanks, I'll take it. I meant "regression" in that it's a test for the future, though!

File: tests/cli/test_main_run.py (hunk 5 above)
Comment (agriyakhetarpal) [reply tests/cli/test_main_run.py:517]: Yes, correct. I hope no one is `conda run` and `conda [de]activate` like this, though. `conda run -- conda activate something` “working” previously would be most likely accidental and not especially meaningful in any robust workflow, as activation would only affect the wrapper’s shell process, not the user’s parent shell, i.e., the user be essentially using an interactive shell mechanism inside a one-shot non-interactive one.

Comment (jaimergp) [reply tests/cli/test_main_run.py:517]: There are workarounds too; they can always eval the hooks directly if they reaaaally need this. I hope no one does. I agree on the accidental component of this working previously.

Comment (ryanskeith) [reply tests/cli/test_main_run.py:517]: Is there a use case for activation like this?  Should we find a way to disable this?

Comment (kenodegard) [reply tests/cli/test_main_run.py:517]: there isn't a good use case, I was just thinking about what this change technically breaks, my only qualm is the error isn't super clear but that's always been an issue with this particular error message

**Comment by agriyakhetarpal** [comment]
Thanks for the approval, @kenodegard! The GitHub project automation moved it to "Approved" as a result – could you please move it back to "Blocked", since we do need to wait for @jezdez's review? Just to make sure this doesn't inadvertently get merged.

File: tests/cli/test_main_run.py (hunk 5 above)
Comment (jezdez) [reply tests/cli/test_main_run.py:517]: Good riddance, I say

**Comment by beckermr** [comment]
@jaimergp @agriyakhetarpal This PR probably needs to be backed out as well. We should roll back to this function as it was in conda 25.11.x and do this all at once with better testing.

File: conda/utils.py
Comment (beckermr) [inline conda/utils.py:1]: @agriyakhetarpal @jaimergp @jezdez This PR combined with the previous one has structural issues IMHO. Before both of those PRs, the various env vars in the hooks that were set, plus the activate process itself was encapsulated in a single spot. This was for good reason since anytime the activation process is changed, it should be changed everywhere. Now with this PR, details specific to activation appear in this utility function. That means if we changed the activation process, we now would need coordinated changes across the code base (i.e., we've lost encapsulation). We need to figure out how to inline things without loosing the encapsulation that was present before.

Comment (agriyakhetarpal) [reply conda/utils.py:1]: Thanks, @beckermr. I'll sit through this a bit more carefully after the 26.3.2 release proceeds. The faulty code is still in `main` (it hasn't been reverted there as noted in #15896). I'll wait until you have results from testing my proposed fix in #15534, and I'll start refactoring the activation code soon after. Thank you for being involved in this manner, and I hope it's okay if I ping you on further work in this direction going forward!

Comment (beckermr) [reply conda/utils.py:1]: Yep. I posted on the other PR the results of the test and it worked. However, the structural issues remain. Do ping me the resulting PR reintroducing these changes and I am more than happy to help review and think through them! 

I really appreciate all of your hard work on conda and your responsiveness on these critical bugs! ❤️
